dandi/utils.py

```diff
@@ -504,3 +509,73 @@ def delayed(*args, **kwargs):
     import joblib
 
     return joblib.delayed(*args, **kwargs)
+
+
+def get_instance(dandi_instance_id):
+    if dandi_instance_id.lower().startswith(("http://", "https://")):
+        redirector_url = dandi_instance_id
+        dandi_id = known_instances_rev.get(redirector_url)
+        if dandi_id is not None:
+            instance = known_instances[dandi_id]
+        else:
+            instance = None
+    else:
+        instance = known_instances[dandi_instance_id]
+        redirector_url = instance.redirector
+        if redirector_url is None:
+            raise ValueError("DANDI instance has no known redirector URL")
+    try:
+        r = requests.get(f"{redirector_url}/server-info")
+        r.raise_for_status()
+    except Exception as e:
+        lgr.warning("Request to %s failed (%s)", redirector_url, str(e))
+        if instance is not None:
+            lgr.warning("Using hard-coded URLs")
+            return instance
+        else:
+            raise RuntimeError(
+                f"Could not retrieve server info from {redirector_url},"
+                " and client does not recognize URL"
+            )
+    server_info = r.json()
+    minversion = Version(server_info["cli-minimal-version"])
+    bad_versions = [Version(v) for v in server_info["cli-bad-versions"]]
+    our_version = Version(__version__)
+    if our_version < minversion:
+        raise CliVersionTooLowError(our_version, minversion, bad_versions)
+    if our_version in bad_versions:
+        raise BadCliVersionError(our_version, minversion, bad_versions)
+    return dandi_instance(
+        girder=server_info["services"]["girder"],
+        gui=server_info["services"]["webui"],
+        redirector=redirector_url,
+    )
+
+
+class CliVersionError(Exception):
```

Comment:
Let's centralize exception definitions in `dandi.exceptions`.  I would probably subclass from `RuntimeError` instead of a generic `Exception`

Reply:
probably worth adding  a docstring (or make `__str__` explicitly an `@abc.abstractmethod`) to note that this is just a base exception, without `__str__` and thus derived classes should be used

Reply:
Done.